### ISSUE TITLE
API SDL_JoystickNumButtons may return invalid number of buttons.

### ISSUE THREAD

**Issue by pktiuk** [libsdl-org/SDL#6442]
For DualSense gamepad:

```
Index:            1
			  UniqueID:         030000004c050000e60c00001181000013563302
			  GUID:             030000004c050000e60c000011810000
			  VendorID:         1356
			  ProductID:        3302
			  Serial:           
			  Product Version:  33041
			  Name:             PS5 Controller
			  Game Controller: Yes
			  # of Axes:       6
			  # of Buttons:    13
			  # of Hats:       0
			  Accelerometer:   0
			  Gyroscope:       0
```

API `SDL_JoystickNumButtons` returned value 13, which is clearly invalid.    
SDL version: 2.24.0

Issue linked with: https://github.com/AntiMicroX/antimicrox/issues/582

**Comment by slouken** [comment]
Why is this invalid? This is being reported via the Linux hid-sony driver, which doesn't expose all the buttons available to the HIDAPI driver. You can debug into src/joystick/linux/SDL_sysjoystick.c to find out exactly what buttons are detected.

**Comment by pktiuk** [comment]
@slouken 
> Why is this invalid?   

Because even when SDL_JoystickNumButtons returns 13 we can still get events from buttons outside of this range (which kinda unexpected)
```cpp
std::cout << "Processing event: " << type << " From joystick with id: " << (int)event.jbutton.which
                    << " Got button with id: " << (int)event.jbutton.button;
```
In some cases gives:
```
Processing event: SDL_CONTROLLERBUTTONUP From joystick with id: 0 Got button with id: 14
```

**Comment by pktiuk** [comment]
In the file `SDL_gamecontroller.h` we can find enum describing selected buttons:

```cpp
typedef enum
{
    SDL_CONTROLLER_BUTTON_INVALID = -1,
    SDL_CONTROLLER_BUTTON_A,
    SDL_CONTROLLER_BUTTON_B,
    SDL_CONTROLLER_BUTTON_X,
    SDL_CONTROLLER_BUTTON_Y,
    SDL_CONTROLLER_BUTTON_BACK,
    SDL_CONTROLLER_BUTTON_GUIDE,
    SDL_CONTROLLER_BUTTON_START,
    SDL_CONTROLLER_BUTTON_LEFTSTICK,
    SDL_CONTROLLER_BUTTON_RIGHTSTICK,
    SDL_CONTROLLER_BUTTON_LEFTSHOULDER,
    SDL_CONTROLLER_BUTTON_RIGHTSHOULDER,
    SDL_CONTROLLER_BUTTON_DPAD_UP,
    SDL_CONTROLLER_BUTTON_DPAD_DOWN,
    SDL_CONTROLLER_BUTTON_DPAD_LEFT,
    SDL_CONTROLLER_BUTTON_DPAD_RIGHT,    //This and all of the above buttons 
    SDL_CONTROLLER_BUTTON_MISC1,    /* Xbox Series X share button, PS5 microphone button, Nintendo Switch Pro capture button, Amazon Luna microphone button */
    SDL_CONTROLLER_BUTTON_PADDLE1,  /* Xbox Elite paddle P1 */
    SDL_CONTROLLER_BUTTON_PADDLE2,  /* Xbox Elite paddle P3 */
    SDL_CONTROLLER_BUTTON_PADDLE3,  /* Xbox Elite paddle P2 */
    SDL_CONTROLLER_BUTTON_PADDLE4,  /* Xbox Elite paddle P4 */
    SDL_CONTROLLER_BUTTON_TOUCHPAD, /* PS4/PS5 touchpad button */
    SDL_CONTROLLER_BUTTON_MAX
} SDL_GameControllerButton;
```

PS5 gamepad has all of the above buttons excluding paddle buttons which gives us total 17 buttons.

**Comment by slouken** [comment]
> Because even when SDL_JoystickNumButtons returns 13 we can still get events from buttons outside of this range (which kinda unexpected)
> 
> ```c++
> std::cout << "Processing event: " << type << " From joystick with id: " << (int)event.jbutton.which
>                     << " Got button with id: " << (int)event.jbutton.button;
> ```
> 
> In some cases gives:
> 
> ```
> Processing event: SDL_CONTROLLERBUTTONUP From joystick with id: 0 Got button with id: 14
> ```

Yes, this is unexpected. Can you debug linux/SDL_sysjoystick.c to see why we're getting a button that we didn't detect?

**Comment by slouken** [comment]
> In the file `SDL_gamecontroller.h` we can find enum describing selected buttons:

The joystick buttons and the game controller buttons are different. There can be any number of joystick buttons and they may or may not map to the set of game controller buttons. Several drivers take advantage of this to expose additional buttons that aren't a part of the game controller spec, but still might be useful to applications.

**Comment by pktiuk** [comment]
> Can you debug linux/SDL_sysjoystick.c to see why we're getting a button that we didn't detect?

Yes

> > In the file `SDL_gamecontroller.h` we can find enum describing selected buttons:
> 
> The joystick buttons and the game controller buttons are different. 

But there is [no other API](https://wiki.libsdl.org/CategoryGameController) which could be used for obtaining number of buttons for the gamepad.

> There can be any number of joystick buttons and they may or may not map to the set of game controller buttons. Several drivers take advantage of this to expose additional buttons that aren't a part of the game controller spec, but still might be useful to applications.

I just showed this enum as a good list of buttons we should expect from a PS5 gamepad.

**Comment by pktiuk** [comment]
> There's no handy "num game controller buttons" API, as you've noted.   

That's a shame.

In this case I think it would be good to have statement clearly mentioned in the docs, because it may be a bit misleading.   
Especially when combined with other statement from [gamecontroller docs](https://wiki.libsdl.org/CategoryGameController):
> This is built on top of the existing joystick API.

**Comment by pktiuk** [comment]
@slouken   

In this case, is there any way of validating whether the button with selected ID exists?   
I tried to use API `SDL_JoystickGetButton`, but in case of calling it with an invalid button ID (like 999), it only returns 0.

**Comment by slouken** [comment]
Yes, you can call `SDL_GameControllerGetBindForButton()` to see if a button is bound.
